KK_ButtPlugin/ButtplugController.cs

```diff
@@ -164,12 +193,49 @@ override protected IEnumerator Run(int girlIndex)
                     minVibration = 0.6f;
                 }
 
+                // allow caress behavior when female vibrations are enabled
+                if (IsFemale && hscene != null && flags.mode == HFlag.EMode.aibu)
+                {
+                    // for now only add touch vibration to any item between the legs
+                    if (hscene.hand.useItems.Any(item => item?.kindTouch == HandCtrl.AibuColliderKind.kokan || item?.kindTouch == HandCtrl.AibuColliderKind.anal))
+                    {
+                        // TODO have different intensities for different items
+                        var vibration = Mathf.InverseLerp(0, flags.speedMaxItem, flags.speedItem) * Mathf.Lerp(0.2f, 1.0f, flags.speedCalc);
+                        DoVibrate(vibration);
+                        yield return new WaitForSecondsRealtime(1.0f / (float)ButtPlugin.VibrationUpdateFrequency.Value);
+                        continue;
+                    }
+                }
+
                 if (ButtPlugin.SyncVibrationWithAnimation.Value)
                 {
                     // Simple sin based intensity amplification based on normalized position in looping animation
                     var info = animator.GetCurrentAnimatorStateInfo(0);
                     var depth = (info.normalizedTime - GetPhase()) % 1;
                     strength = Mathf.Sin(Mathf.Lerp(0, Mathf.PI, depth)) + 0.1f;
+
+
+                    // masturbation is on a non-speed controlled animation
+                    // it has a fixed order of the animation loops, so we can apply a base strength
+                    //   relative to the intensity of the animation
+                    if (IsFemale && flags.mode == HFlag.EMode.masturbation)
+                    {
+                        if (flags.nowAnimStateName == "WLoop")
+                        {
+                            speed = 0.4f;
+                            minVibration = 0.2f;
+                        }
+                        else if (flags.nowAnimStateName == "M_Loop")

```

Comment:
I've never seen M_Loop with an underscore before. You sure it's not MLoop?

Reply:
my bad, I was typing too fast